3.8/graphs-custom-pregel.md

```diff
@@ -0,0 +1,1550 @@
+
+**Advanced example:**
+
+Calculate the sum of all available object values
+
+```js
+["reduce",
+  {"a": 1, "b": 2, "c": 3},
+  ["lambda",
+    ["quote", []],
+    ["quote", ["key", "value", "accum" ]],
+    ["seq",
+      ["quote",
+        [
+          "attrib-set",
+          ["var-ref", "accum"],
+          ["var-ref", "key"],
+          ["+", ["var-ref", "value"], ["attrib-ref", ["var-ref", "accum"], ["var-ref", "key"]] ]
+        ]
+      ]
+    ]
+  ],
+  {"a": 1, "b": 2, "c": 3, "d": 4}
+]
+```
+
+Will produce:
+```
+=> {"a":2, "b":4, "c":6, "d":4}
+```
+
+#### Utilities
+
+_random functions that fit no other category_
+
+```js
+["string-cat", strings...]
+["int-to-str", int]
+["min", numbers...]
+["max", numbers...]
+["avg", numbers...]
+["rand"]
+["rand-range", min, max]
+```
+
+`string-cat` concatenates the given strings. `int-to-string` converts a integer
+to its decimal representation.
+`min`/`max`/`avg` computes the minimum/maximum/average of its values.
+`rand`/`rand-range` produces a pseudo random number uniformly distributed in `[0,1]`/`[min,max]`.
+
+```js
+> ["string-cat", "hello", " ", "world"]
+ = "hello world"
+> ["min", 1, 2, 3]
+ = 1
+> ["max", 1, 2, 3]
+ = 3
+> ["avg", 1, 2, 3]
+ = 2
+> ["rand"]
+ = 0.8401877171547095
+> ["rand-range", 5, 7]
+ = 5.788765853638186
+```
+
+#### Functional
+
+```js
+["id", value]
+["apply", func, list]
+["map", func, list]
+["map", func, dict]
+["filter", func, list]
+["filter", func, dict]
+```
+
+`id` returns its argument. `apply` invokes `func` using the values from `list`
+as arguments. `map` invokes `func` for every value/key-value-pair in the
+`list`/`dict`. `func` should accept two parameters `(index, value)`/`(key, value)`.
+`filter` returns a new `list`/`dict` that contains all entries for which the return value
+of `func` invoked with `(index, value)`/`(key, value)` is considered true.
+
+```js
+> ["id", 12]
+ = 12
+> ["apply", "min", ["quote", 1, 2, 3]]
+ = 1
+> ["map", ["lambda", ["list"], ["list", "idx", "value"], ["quote", ["int-to-str", ["var-ref", "value"]]]], ["list", 1, 2, 3, 4]]
+ = ["1", "2", "3", "4"]
+> ["filter", ["lambda",
+        ["list"],
+        ["list", "idx", "value"],
+        ["quote", ["gt?", ["var-ref", "value"], 3]]
+      ], ["list", 1, 2, 3, 4, 5, 6]]
+ = [4,5,6]
+```
+
+#### Variables
+
+```js
+["var-ref", name]
+["bind-ref", name]
+```
+
+`var-ref` evaluates to the current value of the variable with name `name`.
+It is an error to reference a variable that is not defined in the current
+context. `bind-ref` is an alias of `var-ref`.
+
+#### Debug operators
+
+```js
+["print", values...]
+["error", msg...]
+["assert", cond, msg...]
+```
+
+`print` print in a context dependent way the string representation of its
+arguments joined by spaces. Strings represent themselves, numbers are converted
+to decimal representation, booleans are represented as `true` or `false`.
+Dicts and lists are converted to JSON.
```

Comment:
I created a ticket. For now I changed the name to "report" and it will generate an error, containing the message to be printed and a message that explains that printing is not supported, when no print callback is set. We never write to `stderr`.